dev/io.openliberty.org.jboss.resteasy.common/src/org/jboss/resteasy/core/ServerResponseWriter.java

```diff
@@ -0,0 +1,752 @@
+               }
+            }
+         }
+      }
+      return mt;
+   }
+
+   private static void executeFilters(BuiltResponse jaxrsResponse, HttpRequest request, HttpResponse response,
+         ResteasyProviderFactory providerFactory,
+         ResourceMethodInvoker method, Consumer<Throwable> onComplete, RunnableWithIOException continuation) throws IOException
+   {
+      ContainerResponseFilter[] responseFilters = null;
+
+      if (method != null)
+      {
+         responseFilters = method.getResponseFilters();
+      }
+      else
+      {
+         responseFilters = providerFactory.getContainerResponseFilterRegistry().postMatch(null, null);
+      }
+
+      if (responseFilters != null)
+      {
+         ResponseContainerRequestContext requestContext = new ResponseContainerRequestContext(request);
+         ContainerResponseContextImpl responseContext = new ContainerResponseContextImpl(request, response, jaxrsResponse,
+               requestContext, responseFilters, onComplete, continuation);
+
+         RESTEasyTracingLogger logger = RESTEasyTracingLogger.getInstance(request);
+
+         final long timestamp = logger.timestamp("RESPONSE_FILTER_SUMMARY");
+         // filter calls the continuation
+         responseContext.filter();
+         logger.logDuration("RESPONSE_FILTER_SUMMARY", timestamp, responseFilters.length);
+      }
+      else
+      {
+         try
+         {
+            continuation.run(onComplete);
+         }
+         catch(Throwable t)
+         {
+            onComplete.accept(t);
+            SynchronousDispatcher.rethrow(t);
+         }
+      }
+   }
+
+   protected static void setDefaultContentType(HttpRequest request, BuiltResponse jaxrsResponse, ResteasyProviderFactory providerFactory, ResourceMethodInvoker method)
+   {
+      MediaType chosen = getDefaultContentType(request, jaxrsResponse, providerFactory, method);
+      jaxrsResponse.getHeaders().putSingle(HttpHeaders.CONTENT_TYPE, chosen);
+   }
+
+   protected static MediaType defaultTextPlain = null;
+   
+   @SuppressWarnings("rawtypes")
+   protected static MediaType getDefaultContentType(HttpRequest request, BuiltResponse jaxrsResponse, ResteasyProviderFactory providerFactory, ResourceMethodInvoker method)
+   {
+      // Note. If we get here before the request is executed, e.g., if a ContainerRequestFilter aborts,
+      // chosen and method can be null.
+
+      MediaType chosen = (MediaType)request.getAttribute(SegmentNode.RESTEASY_CHOSEN_ACCEPT);
+      
+   // Liberty change begin
+      // if the type/subtype matches a static MediaType from the API, return the static
+      if((chosen != null) && isConcrete(chosen)) {
+          String key = chosen.getType().trim() + "/" + chosen.getSubtype().trim();
+          MediaType cached = mediaTypeCache.get(key);
+          if(cached != null) {
+              return cached;
+          }
+      }
```

Comment:
This doesn't work in the !hasProduces case.  I don't understand the point of having a Map<MediaType,MediaType>.  It could just be a Set<MediaType> and return chosen instead of getting the statically defined one.  It would seem that is the class or method has `@Produces` on it most everything gets skipped and won't be an issue.  The only thing a cache saves on is the containsKey at the end of the class.

Reply:
> I don't understand the point of having a Map<MediaType,MediaType>

The initial implementation took care to remove several elements (HAS_PRODUCES, q, qs) from the MediaType parameters map before returning the MediaType, so just returning the MediaType as extracted from the original request seems wrong. 

The point of the cache is to skip the manipulation of the original MediaType object at the end of the method, after the if(hasProduces) clause - that's where most of the time in this method is spent with our simple workload.

- ~1% creating new HashMaps
- ~0.85% creating new MediaTypes
- ~0.6% calling containsKey 
- etc

That's all unnecessary - all we need to do is return the static MediaType corresponding to the type and subtype of the MediaType extracted from the request. 

The initial form of getDefaultContentType with all that manual MediaType crafting takes about 3.5-4.0% of ticks, while the new form with the cache has that down to ~0.6-0.7%, for the primitive ping workload used in our measurement.